### PR TITLE
syncer: add table router and binlog event filters in drainer

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
To implement sync to oracle drainer need to implement binlog event filters and table router first.

### What is changed and how it works?
All the configurations are implemented in dm v2 configuration templates.
1. Implement binlog event filters and skip steps in `drainer.syncer`.
2. Implement table router but didn't add it to any `dsyncer` except for oracle dsyncer. After a second thought we think that router configuration has some danger for synchronization. We will open this configuration to other dsyncers after a full GA process.
3. This PR doesn't use `table-migrate-rule` to generate baList. We can take a further discussion in this PR later.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### Release note

 - No release note

### CODE REVIEW

**Comment by lichunzhu** [comment]
/rebuild
will add some tests later

**Comment by lichunzhu** [comment]
@glorv @lance6716 PTAL

File: drainer/syncer.go [hunk 1]
```diff
@@ -550,6 +582,66 @@ func filterTable(pv *pb.PrewriteValue, filter *filter.Filter, schema *Schema) (i
...
+				default:
+					err = errors.Errorf("unknown mutation type: %v", tp)
+					return
+				}
+
+				needSkip, err := skipByFilter(binlogFilter, schemaName, tableName, et, "")
+				if err != nil {
+					return false, errors.Trace(err)
+				}
+				if !needSkip {
```
Comment (lance6716) [inline drainer/syncer.go:617]: seems we can

if needSkip {continue}

File: tests/binlogfilter/drainer.toml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+[syncer.binlog-filter-rule]
+
+[syncer.binlog-filter-rule.truncate-table-filter]
+ignore-event = ["truncate table"]
+ignore-sql = []
+
+[syncer.binlog-filter-rule.add-column-aaa-filter]
+ignore-event = []
+ignore-sql = ["alter table .* add column aaa int"]
```
Comment (lance6716) [inline tests/binlogfilter/drainer.toml:55]: does TiDB binlog wrote DDL binlog after normalization? (\`aaa\`)

**Review by lance6716 — APPROVED** [review]

File: tests/binlogfilter/drainer.toml (hunk 2 above)
Comment (lichunzhu) [reply tests/binlogfilter/drainer.toml:55]: No. Just the original DDL.

File: drainer/syncer.go [hunk 3]
```diff
@@ -61,6 +66,11 @@ type Syncer struct {
 
 	shutdown chan struct{}
 	closed   chan struct{}
+
+	// TODO: to discuss block allow list implementation in the future
+	// baList       *baf.Filter
```
Comment (glorv) [inline drainer/syncer.go:71]: Does this mean we want to use baList to replace the  do-dbs/tables ignore-dbs/tables?

File: drainer/syncer.go [hunk 4]
```diff
@@ -562,6 +654,37 @@ func filterTable(pv *pb.PrewriteValue, filter *filter.Filter, schema *Schema) (i
...
+func skipDDLEvent(sql, schema, table string, p *parser.Parser, binlogFilter *bf.BinlogEvent) (ignore bool, err error) {
+	if binlogFilter == nil {
+		return false, nil
+	}
+	stmt, err := p.ParseOneStmt(sql, "", "")
+	if err != nil {
+		log.L().Error("fail to parse ddl", zap.String("ddl", sql), logutil.ShortError(err))
+		// return error if parse fail and filter fail
+		needSkip, err2 := skipByFilter(binlogFilter, schema, table, bf.NullEvent, sql)
+		return needSkip, err2
```
Comment (glorv) [inline drainer/syncer.go:668]: IF needSkip is false here, do we need to return a error as this is an unknown statement.

**Review by glorv — COMMENTED** [review]
rest LGTM

File: drainer/syncer.go (hunk 3 above)
Comment (lichunzhu) [reply drainer/syncer.go:71]: I will delete this part. As we discussed before, we will **NOT** use baList to replace the do-dbs/tables ignore-dbs/tables.

File: drainer/syncer.go (hunk 4 above)
Comment (lichunzhu) [reply drainer/syncer.go:668]: I think that doesn't matter. If this can't be executed downstream they will return an error.

**Review by glorv — APPROVED** [review]

File: drainer/syncer.go [hunk 5]
```diff
@@ -550,6 +580,66 @@ func filterTable(pv *pb.PrewriteValue, filter *filter.Filter, schema *Schema) (i
...
+			mutation.Sequence = mutation.Sequence[0:filteredIdx]
+			if mutation.InsertedRows != nil {
+				mutation.InsertedRows = mutation.InsertedRows[0:filteredInsertIdx]
+			}
+			if mutation.UpdatedRows != nil {
+				mutation.UpdatedRows = mutation.UpdatedRows[0:filteredUpdateIdx]
+			}
+			if mutation.DeletedRows != nil {
+				mutation.DeletedRows = mutation.DeletedRows[0:filteredDeleteIdx]
+			}
```
Comment (glorv) [inline drainer/syncer.go:640]:
```suggestion
			mutation.InsertedRows = mutation.InsertedRows[0:filteredInsertIdx]
			mutation.UpdatedRows = mutation.UpdatedRows[0:filteredUpdateIdx]
			mutation.DeletedRows = mutation.DeletedRows[0:filteredDeleteIdx]
```
If rows is nil, then idx should be 0, so we can still do rows[0:0] safely. 😁

**Pull request merged**